File: NetDeck/AboutViewController.m
//
//  AboutViewController.m
//  Net Deck
//
//  Created by Gereon Steffens on 25.03.13.
//  Copyright © 2016 Gereon Steffens. All rights reserved.
//

#import "AboutViewController.h"

@interface AboutViewController ()
@property MFMailComposeViewController *mailer;
@property UIBarButtonItem* backButton;
@end

@implementation AboutViewController

- (id)init
{
    return [self initWithNibName:@"AboutView" bundle:nil];
}

- (id)initWithNibName:(NSString *)nibNameOrNil bundle:(NSBundle *)nibBundleOrNil
{
    return [super initWithNibName:nibNameOrNil bundle:nibBundleOrNil];
}

-(void) dealloc
{
    self.webView.delegate = nil;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.navigationController.navigationBar.barTintColor = [UIColor whiteColor];
    self.navigationController.navigationBar.topItem.title = l10n(@"About");
    
    self.backButton = [[UIBarButtonItem alloc] initWithTitle:@"◁" style:UIBarButtonItemStylePlain target:self action:@selector(goBack:)];
    
    self.webView.delegate = self;
    self.webView.dataDetectorTypes = UIDataDetectorTypeNone;
        
    NSURL* url= [NSURL fileURLWithPath:[[NSBundle mainBundle] pathForResource:@"About" ofType:@"html"] isDirectory:NO];
    [self.webView loadRequest:[NSURLRequest requestWithURL:url]];
}

-(void) viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    
    NSString* title = [NSString stringWithFormat:l10n(@"About Net Deck %@"), [AppDelegate appVersion]];
    if (Device.isIphone)
    {
        title = [NSString stringWithFormat:l10n(@"Net Deck %@"), [AppDelegate appVersion]];
    }
    
    UINavigationItem* topItem = self.navigationController.navigationBar.topItem;
    topItem.title = title;
    topItem.rightBarButtonItem = [[UIBarButtonItem alloc] initWithTitle:l10n(@"Feedback") style:UIBarButtonItemStylePlain target:self action:@selector(leaveFeedback:)];
}

#pragma mark buttons

-(void) goBack:(id)sender
{
    [self.webView goBack];
    self.navigationController.navigationBar.topItem.leftBarButtonItem = nil;
}

-(void) leaveFeedback:(id)sender
{
    NSString* msg = l10n(@"We'd love to know how we can make Net Deck even better - and would really appreciate if you left a review on the App Store.");
    
    UIAlertController* alert = [UIAlertController alertWithTitle:nil message:msg];
    
    [alert addAction:[UIAlertAction actionWithTitle:l10n(@"Write a Review") handler:^(UIAlertAction * action) {
        [self rateApp];
    }]];
    [alert addAction:[UIAlertAction actionWithTitle:l10n(@"Contact Developers") handler:^(UIAlertAction * action) {
        [self sendEmail];
    }]];
    [alert addAction:[UIAlertAction alertCancel:nil]];

    [self presentViewController:alert animated:NO completion:nil];
}

-(void) rateApp
{
    NSString* appURL = [NSString stringWithFormat:@"itms-apps://itunes.apple.com/%@/app/id%@",
                        [[NSLocale preferredLanguages] objectAtIndex:0],
                        @"865963530"];
    
    [[UIApplication sharedApplication] openURL:[NSURL URLWithString:appURL]];
}

-(void) sendEmail
{
    self.mailer = [[MFMailComposeViewController alloc] init];
    
    if (self.mailer) // see crashlytics #57
    {
        self.mailer.mailComposeDelegate = self;
        [self.mailer setToRecipients:@[ @"netdeck@steffens.org" ]];
        NSMutableString* subject = [NSMutableString stringWithString:l10n(@"Net Deck Feedback ")];
        [subject appendString:[AppDelegate appVersion]];
        [self.mailer setSubject:subject];
        [self presentViewController:self.mailer animated:NO completion:nil];
    }
}

#pragma mark mail compose

-(void)mailComposeController:(MFMailComposeViewController*)controller didFinishWithResult:(MFMailComposeResult)result error:(NSError*)error
{
    [self.mailer dismissViewControllerAnimated:NO completion:nil];
    self.mailer = nil;
}

#pragma mark webview

-(BOOL) webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest*)request navigationType:(UIWebViewNavigationType)type
{
    if (type == UIWebViewNavigationTypeLinkClicked)
    {
        NSString* scheme = [request URL].scheme;
        if ([scheme isEqualToString:@"mailto"])
        {
            [self sendEmail];
        }
        else if ([scheme isEqualToString:@"itms-apps"])
        {
            [self rateApp];
        }
        else if ([scheme isEqualToString:@"file"])
        {
            NSString* path = [[NSBundle mainBundle] pathForResource:@"Acknowledgements" ofType:@"html"];
            
            NSURL* url = [NSURL fileURLWithPath:path isDirectory:NO];
            [self.webView loadRequest:[NSURLRequest requestWithURL:url]];
            self.navigationController.navigationBar.topItem.leftBarButtonItem = self.backButton;
            return YES;
        }
        else
        {
            [[UIApplication sharedApplication] openURL:[request URL]];
        }
        return NO;
    }
    
    return YES;
}

@end

File: NetDeck/AboutViewController.swift
//
//  AboutViewController.swift
//  NetDeck
//
//  Created by Gereon Steffens on 05.01.17.
//  Copyright © 2019 Gereon Steffens. All rights reserved.
//

import UIKit
import MessageUI

@objc(AboutViewController) // needed for IASK
class AboutViewController: UIViewController, UIWebViewDelegate, MFMailComposeViewControllerDelegate {
    
    @IBOutlet weak var webView: UIWebView!
    
    private var backButton: UIBarButtonItem!
    private var mailer: MFMailComposeViewController!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.backButton = UIBarButtonItem(title: "◁", style: .plain, target: self, action: #selector(self.goBack(_:)))
        
        self.webView.delegate = self
        self.webView.dataDetectorTypes = []
        
        if let path = Bundle.main.path(forResource: "About", ofType: "html") {
            let url = URL(fileURLWithPath: path)
            self.webView.loadRequest(URLRequest(url: url))
        }
    }
    
    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        let fmt = Device.isIphone ? "Net Deck %@" : "About Net Deck %@"
        let title = String(format: fmt.localized(), Utils.appVersion())
        
        self.title = title
        
        let feedbackButton = UIBarButtonItem(title: "Feedback".localized(), style: .plain, target: self, action: #selector(self.leaveFeedback(_:)))
        self.navigationItem.rightBarButtonItem = feedbackButton

        Analytics.logEvent(.showAbout)
    }
    
    @objc func goBack(_ sender: Any) {
        self.webView.goBack()
        self.navigationItem.leftBarButtonItem = nil
    }
    
    @objc func leaveFeedback(_ sender: Any) {
        let msg = "We'd love to know how we can make Net Deck even better - and would really appreciate if you left a review on the App Store.".localized()
        
        let alert = UIAlertController(title: nil, message: msg, preferredStyle: .alert)
        alert.addAction(UIAlertAction(title: "Write a Review".localized()) { action in
            self.rateApp()
        })
        alert.addAction(UIAlertAction(title: "Contact Developers".localized()) { action in
            self.sendEmail()
        })
        alert.addAction(UIAlertAction.alertCancel(nil))
        
        self.present(alert, animated: false, completion: nil)
        
    }
    
    func rateApp() {
        if let url = URL(string: "itms-apps://itunes.apple.com/app/id865963530") {
            UIApplication.shared.open(url)
        }
    }
    
    func sendEmail() {
        guard MFMailComposeViewController.canSendMail() else {
            return
        }
        self.mailer = MFMailComposeViewController()
        if self.mailer != nil {
            self.mailer.mailComposeDelegate = self
            self.mailer.setToRecipients([ "netdeck@steffens.org" ])
            let subject = "Net Deck Feedback ".localized() + Utils.appVersion()
            self.mailer.setSubject(subject)
            self.present(self.mailer, animated: false, completion: nil)
        }
    }
    
    func mailComposeController(_ controller: MFMailComposeViewController, didFinishWith result: MFMailComposeResult, error: Error?) {
        self.mailer.dismiss(animated: false, completion: nil)
        self.mailer = nil
    }
    
    func webView(_ webView: UIWebView, shouldStartLoadWith request: URLRequest, navigationType: UIWebView.NavigationType) -> Bool {
        
        if navigationType == .linkClicked {
            let scheme = request.url?.scheme ?? ""
            
            switch scheme {
            case "mailto":
                self.sendEmail()
            case "itms-apps":
                self.rateApp()
            case "file":
                if let path = Bundle.main.path(forResource: "Acknowledgements", ofType: "html") {
                    let url = URL(fileURLWithPath: path)
                    self.webView.loadRequest(URLRequest(url: url))
                    self.navigationItem.leftBarButtonItem = self.backButton
                    return true
                }
            default:
                if let url = request.url {
                    UIApplication.shared.open(url)
                }
            }
            return false
        }
        return true
    }
}
